File: app/src/main/java/com/raredev/vcspace/ui/editor/EditorViewModel.java
package com.raredev.vcspace.ui.editor;

import androidx.core.util.Pair;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import java.io.File;
import java.util.ArrayList;
import java.util.List;

public class EditorViewModel extends ViewModel {
  private final MutableLiveData<Pair<Integer, File>> currentPosition = new MutableLiveData<>(new Pair(-1, null));

  private MutableLiveData<List<File>> mFiles = new MutableLiveData<>(new ArrayList<>());

  public LiveData<Pair<Integer, File>> getCurrentPositionPair() {
    return currentPosition;
  }
  
  public int getCurrentPosition() {
    return currentPosition.getValue().first;
  }

  public File getCurrentFile() {
    return currentPosition.getValue().second;
  }
  
  public void setCurrentPosition(int pos, File file) {
    currentPosition.setValue(new Pair(pos, file));
  }

  public LiveData<List<File>> getFiles() {
    return mFiles;
  }

  public void clear() {
    List<File> value = getFiles().getValue();

    value.clear();
    mFiles.setValue(value);
    setCurrentPosition(-1, null);
  }

  public void addFile(File file) {
    List<File> files = getFiles().getValue();

    files.add(file);
    mFiles.setValue(files);
  }

  public void removeFile(int index) {
    List<File> files = getFiles().getValue();

    files.remove(index);
    
    if (files.isEmpty()) {
      setCurrentPosition(-1, null);
    }
    mFiles.setValue(files);
  }

  public boolean contains(Object obj) {
    List<File> files = getFiles().getValue();
    if (files.isEmpty()) {
      return false;
    }
    return files.contains(obj);
  }

  public int indexOf(Object obj) {
    return getFiles().getValue().indexOf(obj);
  }
}

File: app/src/main/java/com/raredev/vcspace/viewmodel/EditorViewModel.kt
package com.raredev.vcspace.viewmodel

import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModel
import java.io.File

class EditorViewModel : ViewModel() {

  private val _files = MutableLiveData<MutableList<File>>(mutableListOf())
  private val _selectedFile = MutableLiveData<Pair<Int, File?>>(-1 to null)

  val openedFiles: List<File>
    get() = _files.value!!

  val fileCount: Int
    get() = _files.value!!.size

  val selectedFile: File?
    get() = _selectedFile.value!!.second

  val selectedFileIndex: Int
    get() = _selectedFile.value!!.first

  fun addFile(file: File) {
    val files = this._files.value!!
    files.add(file)
    this._files.value = files
  }

  fun updateFile(index: Int, file: File) {
    val files = this._files.value!!
    files[index] = file
    this._files.value = files
  }

  fun removeFile(index: Int) {
    val files = this._files.value!!
    files.removeAt(index)
    this._files.value = files

    if (files.isEmpty()) setSelectedFile(-1, null)
  }

  fun removeAllFiles() {
    val files = this._files.value!!
    files.clear()
    this._files.value = files
    setSelectedFile(-1, null)
  }

  fun observeFiles(lifecycleOwner: LifecycleOwner, observer: Observer<MutableList<File>>) {
    this._files.observe(lifecycleOwner, observer)
  }

  fun setSelectedFile(index: Int) {
    setSelectedFile(index, openedFiles[index])
  }

  fun setSelectedFile(index: Int, file: File?) {
    this._selectedFile.value = index to file
  }

  fun observeSelectedFile(lifecycleOwner: LifecycleOwner, observer: Observer<Pair<Int, File?>>) {
    this._selectedFile.observe(lifecycleOwner, observer)
  }
}
